src/storage/src/source/antichain.rs

```diff
@@ -0,0 +1,165 @@
+// by the Apache License, Version 2.0.
+
+//! Drivers for upstream commit
+use std::collections::HashMap;
+
+use mz_expr::PartitionId;
+
+use crate::types::sources::MzOffset;
+
+/// OffsetAntichain is similar to a timely `Antichain<(PartitionId, T: TotalOrder)>`,
+/// but additionally:
+///
+/// - Uses a HashMap as the implementation to allow absence of a `PartitionId` to mean
+/// that `PartitionId` is at `T::minimum`. This helps avoid needing to hold onto a HUGE
+/// `Antichain` for all possible `PartitionId`s
+///     - Note this means that a partition being "finished" (like a normal "empty"
+///     `Antichain`, is not currently supported, but could be added
+///     - Note that this `Antichain` can also have been filtered, as in, missing some
+///     partitions for which data exists but we don't care about. This is semantically
+///     different than if we just don't have data, but it is represented the same
+/// - Is not generic over `T`, but instead uses `MzOffset`, which we have
+/// - Allows users to go from a _frontier_ to an actual set of offsets that are
+/// connected to real data.
+///     - This is a consequence of implementation, where the _frontier_ is ALWAYS
+///     generated from real data offsets, in an invertible way.
+///
+/// `OffsetAntichain` has 4 sets of Api's:
+/// - "read" apis like `get` and `as_vec`
+/// - "mutation" apis (currently only `filter_by_partition`)
+/// - And 2 "write" apis, that should primarily be used separately from each other:
+///   - "Frontier" apis, which directly manipulate the underlying frontier.
+///   Useful for implementing primitives like reclocking
+///   - "Data" apis, that maintain special invariants:
+///     - `insert_data_up_to` updates the frontier based a given offset
+///     that
+///     - `as_data_offsets` inverts the behavior of `insert_data_up_to`
+///     and returns a `HashMap<PartitionId, MzOffset>` of offets
+///     of real committed data.
+#[derive(Debug, PartialEq, Clone, serde::Serialize, serde::Deserialize)]
+pub struct OffsetAntichain {
+    inner: HashMap<PartitionId, MzOffset>,
+}
+
+impl PartialEq<HashMap<PartitionId, MzOffset>> for OffsetAntichain {
+    fn eq(&self, other: &HashMap<PartitionId, MzOffset>) -> bool {
+        other == &self.inner
+    }
+}
+
+impl OffsetAntichain {
+    /// Initialize an Antichain where all partitions have made no progress.
+    pub fn new() -> Self {
+        Self {
+            inner: HashMap::new(),
+        }
+    }
+
+    /// Initialize an Antichain where all partitions have made no progress,
+    /// but with `cap` capacity in the underlying data structure.
+    pub fn with_capacity(cap: usize) -> Self {
+        Self {
+            inner: HashMap::with_capacity(cap),
+        }
+    }
+
+    // Data apis
+
+    /// Advance the frontier represented by this `OffsetAntichain` for `pid`
+    /// to a value that contains `offset`.
+    pub fn insert_data_up_to(&mut self, pid: PartitionId, offset: MzOffset) {
```

Comment:
why did you go with these "data" methods instead of requiring callers to to the `+1` logic?

Reply:
I wanted the logic that +1's and -1's to be scoped to a single module!